ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,412 @@
+end
+```
+
+
+## Errors
+
+Here's the Error hierarchy that is planned for the Ruby SDK:
+
+```ruby
+module Temporal
+  # top level error superclass
+  class Error < StandardError
+
+  # superclass for all errors within the SDK itself
+  class InternalError < Error
+
+  # errors specific to Client
+  class ClientError < Error
+
+  # errors specific to Worker
+  class WorkerError < InternalError
+
+  # superclass for connection, network and response errors
+  class RPCError < Error
+  class UnexpectedError < RPCError
+
+  # superclass for all API failure responses
+  class APIError < ServerError
+
+  # type hinting errors
+  class TypeError < Error
+
+  # superclass for workflow errors
+  class WorkflowError < Error
+```
+
+Notes:
+
+- This list is not exhaustive, more errors will be added as parts of the SDK are getting built
+- Some concrete error class placement might be confusing and will be clarified later in the process
+
+
+### Converting API Failures
+
+Similar to other SDKs we will decode `Failure` proto messages to native error classes. All `cause`s
+will be decoded recursively and the result will form a hierarchy identical to the one in the proto
+response. These error classes will all be subclasses of an `APIError` and will have names matching
+the original protobuf messages (while replacing `Failure` with `Error` for consistency within Ruby):
+
+```ruby
+module Temporal
+  class ApplicationError < APIError
+  class TimeoutError < APIError
+  class CanceledError < APIError
+  class TerminatedError < APIError
+  class ServerError < APIError
+  class ResetWorkflowError < APIError
+  class ActivityError < APIError
+  class ChildWorkflowExecutionError < APIError
+end
+```
+
+At this point we are not going to be converting user errors back to their original classes. These
+will be decoded into `ApplicationError`s while keeping the original class name, message, stack trace
+and other details. Here's an example of handling an activity error:
+
+```ruby
+class ProcessPaymentActivity < Temporal::Activity
+  class InsufficientFunds < Temporal::NonRetryableError; end
+  class PaymentNetworkError < Temporal::RetryableError
+    attr_reader :status
+
+    def initialize(message, status)
+      super(message)
+      @status = status
+    end
+  end
+
+  def execute(account, amount)
+    if amount < balance_of(account.number)
+      raise InsufficientFunds, "Account #{account.number} does not have enough funds"
+    end
+
+    response = process_payment(account.number, amount)
+    if response.failure?
+      raise PaymentNetworkError, response.message, response.status
+    end
+
+    ...
+  end
+end
+
+class RenewSubscriptionWorkflow < Temporal::Workflow
+  def execute
+    ...
+
+    begin
+      workflow.execute_activity(ProcessPaymentActivity, account, amount)
+    rescue Temporal::ActivityError => error
+      if error.cause.is_a?(Temporal::ApplicationError)
+        case error.cause.original_class
```

Comment:
"original_class" isn't something available on our failure proto. Where is this field populated from? If it's "type" we should probably keep that name because other languages use it too and "class" is a Ruby concept.

Reply:
It is the `type`, however since this is `Temporal::ApplicationError` (and not a protobuf Failure) I think we can rename it to provide a bit more context. `Type` doesn't make much sense within Ruby.

I can keep the `type` available as well in case you feel strongly about it

Reply:
I support other names if `type` is unclear. But `original_class` is clearly a Ruby-ism and incorrect when an activity failed somewhere else that just set the type string.

Reply:
I'd stay close to the protos and the other SDKs here, the protos define the field as "type".